src/lib/core/prbuf.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "prbuf.h"
+#include "bit/bit.h"
+
+/**
+ * Partitioned ring buffer. Each entry stores size before user data.
+ * So the typical buffer looks like:
+ * HEADER uint32 DATA uint32 DATA ...
+ *
+ * We have to store offsets to be able to restore buffer (including
+ * all metadata) from raw pointer. Otherwise it is impossible to point
+ * out where head/tail are located.
+ */
+struct PACKED prbuf_header {
+	/**
+	 * Buffer's data layout can be changed in the future, so for the sake
+	 * of proper recovery of the buffer we store its version.
+	 */
+	uint32_t version;
+	/** Total size of buffer (including header). */
+	uint32_t size;
+	/**
+	 * Offset of the oldest entry - it is the first candidate to be
+	 * overwritten. Note that in contrast to iterator/entry - this offset
+	 * is calculated to the first byte of entry (i.e. header containing
+	 * size of entry).
+	 * */
+	uint32_t offset_begin;
+	/** Offset of the next byte after the last (written) record. */
+	uint32_t offset_end;
+};
+
+/**
+ * Structure representing record stored in the buffer so it has the same
+ * memory layout.
+ */
+struct prbuf_record {
+	/** Size of data. */
+	uint32_t size;
+	/** Data. */
+	char data[];
+};
+
+/**
+ * Current prbuf implementation version. Must be bumped in case storage
+ * format is changed.
+ */
+static const uint32_t prbuf_version = 0;
+
+/**
+ * prbuf is assumed to store all metadata in little-endian format. Beware
+ * when decoding its content. Size overhead per store operation is 4 bytes;
+ * moreover not the whole space of the buffer can be used since we do not
+ * wrap entry if it doesn't fit til the buffer's end.
+ *
+ * There are several assumptions regarding the buffer:
+ * - Buffer always contains at least one element;
```

Comment:
It was made in this way to avoid extra check of emptiness in prbuf_prepare().
Reworked: now at the start there's no entries; introduced prbuf_is_empty() function.